### ISSUE TITLE
Seq2Seq原理到实现

### ISSUE THREAD

**Issue by superjcd** [superjcd/comments#1]
https://superjcd.github.io/p/seq2seq%E5%8E%9F%E7%90%86%E5%88%B0%E5%AE%9E%E7%8E%B0/ 

 Seq2Seq 模型架构 seq2seq是2014年随论文[Sequence to Sequence Learning with Neural Networks]发布 的一种NLG自然语言（生成式）模型， 下图直观展示了seq2seq模型在机器翻译任务（德译英）中，神经网络前向计算（Forwad Computiation）的工作流程
seq2seq模型的关键点在于， 将模型分成了编码器和解码器：
编码器在编码阶段，会对每一个Token（上面的话是一个个的单词）进行编码，编码过程通常会使用RNN来实现(论文里使用的是LSTM， 所以会同时有hidden state和cell两个输出， 如果是最原始的RNN， 那么应该就只有一个hidden state)。
解码器会沿用编码器的结构， 比如这里的编码器用的是双层的LSTM(论文里面是4层)， 那么解码器也会使用双层的LSTM， 同时解码器的最初的hidden state和cell刚好就是解码器的最后一步产出的hidden state和cell
hidden state的维度和cell的维度通常会一样
编码器 我们拆解一下整个seq2seq， 先只关注编码器这个部分
编码器前向计算涉及的公式：
e指的是embedding词嵌入
也就是当我们使用一般的RNN模型时， 就是上面的公式， 如果是LSTM, 因为会涉及cell（记忆单元）所以会多一个参数。上面的公式没有涉及层级， 如果涉及多个层级（只有第一层会用到embedding）， 以LSTM模型为例， 公式为：
从第二层layer开始， 编码器的第一个参数变成了前一层的hidden state而不是embbeding
编码器pytorch实现 1 2 3 4 5 6 7 8 9 10 11 12 13 class Encoder(nn.Module): def __init__(self, input_dim, embedding_dim, hidden_dim, n_layers, dropout): super().__init__() self.hidden_dim = hidden_dim self.

**Comment by superjcd** [comment]
Just a test comment
